File: src/main/java/be/breina/openrgb/model/Device.java
package be.breina.openrgb.model;

import static lombok.AccessLevel.NONE;

import java.io.IOException;
import java.util.function.Consumer;
import lombok.Getter;
import lombok.ToString;
import be.breina.openrgb.util.Constants;
import be.breina.openrgb.util.LittleEndianInputStream;

@ToString
@Getter
public class Device {

    private final DeviceType type;

    private final String name;
    private String vendor;
    private final String description, version, serial, location;

    @Getter(NONE)
    private final int activeModeIndex;
    private final Mode[] modes;

    private final Zone[] zones;
    private final Led[] leds;
    private final Color[] colors;

    @Getter(NONE)
    private Consumer<java.awt.Color[]> colorSetter;

    public Mode getActiveMode() {
        return modes[activeModeIndex];
    }

    Device(LittleEndianInputStream input) throws IOException {
        final var duplicatePacketLength = input.readInt();
        type = DeviceType.values()[input.readInt()];
        name = input.readAscii();
        if (Constants.serverProtocol >= 1) {
            vendor = input.readAscii();
        }
        description = input.readAscii();
        version = input.readAscii();
        serial = input.readAscii();
        location = input.readAscii();

        final var modeCount = input.readUnsignedShort();
        modes = new Mode[modeCount];
        activeModeIndex = input.readInt();
        for (int i = 0; i < modeCount; i++) {
            modes[i] = new Mode(input);
        }

        final var zoneCount = input.readUnsignedShort();
        zones = new Zone[zoneCount];
        for (int i = 0; i < zoneCount; i++) {
            zones[i] = new Zone(input);
        }

        final var ledCount = input.readUnsignedShort();
        leds = new Led[ledCount];
        for (int i = 0; i < ledCount; i++) {
            leds[i] = new Led(input);
        }

        final var colorCount = input.readUnsignedShort();
        colors = new Color[colorCount];
        for (int i = 0; i < colorCount; i++) {
            colors[i] = new Color(input);
        }
    }

    public void setColors(java.awt.Color... colors) {
        assert colors.length <= this.colors.length;
        colorSetter.accept(colors);
    }

    void colorMutator(Consumer<java.awt.Color[]> colorSetter) {
        this.colorSetter = colorSetter;
    }
}

File: src/main/java/be/breina/openrgb/model/Device.kt
package be.breina.openrgb.model

import be.breina.openrgb.model.enumeration.DeviceType
import be.breina.openrgb.util.Constants
import be.breina.openrgb.util.LittleEndianInputStream
import lombok.EqualsAndHashCode
import java.util.function.Consumer

@Suppress("MemberVisibilityCanBePrivate")
class Device private constructor(
    val type: DeviceType,
    val name: String,
    var vendor: String?,
    val description: String,
    val version: String,
    val serial: String,
    val location: String,

    private val activeModeIndex: Int,

    val modes: Array<Mode>,
    val zones: Array<Zone>,
    val leds: Array<Led>,
    val colors: Array<Color>,
) {

    val activeMode: Mode get() = modes[activeModeIndex]

    @EqualsAndHashCode.Exclude
    private var colorSetter: Consumer<Array<java.awt.Color>>? = null

    fun setColors(vararg colors: java.awt.Color) {
        assert(colors.size <= this.colors.size)
        colorSetter!!.accept(arrayOf(*colors))
    }

    internal fun colorMutator(colorSetter: Consumer<Array<java.awt.Color>>?) {
        this.colorSetter = colorSetter
    }

    internal companion object {
        fun from(input: LittleEndianInputStream): Device {
            val duplicatePacketLength = input.readInt()
            val type = DeviceType.values()[input.readInt()]
            val name = input.readAscii()

            val vendor: String? =
                if (Constants.serverProtocol >= 1) {
                    input.readAscii()
                } else {
                    null
                }

            val description = input.readAscii()
            val version = input.readAscii()
            val serial = input.readAscii()
            val location = input.readAscii()

            val modeCount = input.readUnsignedShort()
            val activeModeIndex = input.readInt()
            val modes: Array<Mode> = Array(modeCount) { Mode.from(input) }

            val zoneCount = input.readUnsignedShort()
            val zones: Array<Zone> = Array(zoneCount) { Zone.from(input) }

            val ledCount = input.readUnsignedShort()
            val leds: Array<Led> = Array(ledCount) { Led.from(input) }

            val colorCount = input.readUnsignedShort()
            val colors: Array<Color> = Array(colorCount) { Color.from(input) }

            return Device(type, name, vendor, description, version, serial, location, activeModeIndex, modes, zones, leds, colors)
        }
    }
}